File: app/src/androidTest/java/org/tasks/jobs/NotificationQueueTest.java
package org.tasks.jobs;

import static com.google.common.collect.Sets.newHashSet;
import static com.todoroo.astrid.reminders.ReminderService.TYPE_DUE;
import static com.todoroo.astrid.reminders.ReminderService.TYPE_SNOOZE;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static junit.framework.Assert.assertEquals;
import static org.mockito.AdditionalAnswers.returnsFirstArg;
import static org.mockito.Matchers.anyLong;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;

import android.support.test.runner.AndroidJUnit4;
import java.util.List;
import java.util.concurrent.TimeUnit;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.tasks.Freeze;
import org.tasks.Snippet;
import org.tasks.preferences.Preferences;
import org.tasks.time.DateTime;

@RunWith(AndroidJUnit4.class)
public class NotificationQueueTest {

  private static final long ONE_MINUTE = TimeUnit.MINUTES.toMillis(1);

  private NotificationQueue queue;
  private WorkManager workManager;
  private Preferences preferences;

  @Before
  public void before() {
    preferences = mock(Preferences.class);
    when(preferences.adjustForQuietHours(anyLong())).then(returnsFirstArg());
    workManager = mock(WorkManager.class);
    queue = new NotificationQueue(preferences, workManager);
  }

  @After
  public void after() {
    verifyNoMoreInteractions(workManager);
  }

  @Test
  public void alarmAndReminderSameTimeSameID() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new AlarmEntry(1, 1, now));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(
                    newHashSet(new AlarmEntry(1, 1, now), new ReminderEntry(1, now, TYPE_DUE)),
                    newHashSet(queue.getOverdueJobs()));
              }
            });
  }

  @Test
  public void removeAlarmLeaveReminder() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new AlarmEntry(1, 1, now));

    verify(workManager).scheduleNotification(now, true);

    queue.remove(singletonList(new AlarmEntry(1, 1, now)));

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(
                    singletonList(new ReminderEntry(1, now, TYPE_DUE)), queue.getOverdueJobs());
              }
            });
  }

  @Test
  public void removeReminderLeaveAlarm() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new AlarmEntry(1, 1, now));

    verify(workManager).scheduleNotification(now, true);

    queue.remove(singletonList(new ReminderEntry(1, now, TYPE_DUE)));

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(singletonList(new AlarmEntry(1, 1, now)), queue.getOverdueJobs());
              }
            });
  }

  @Test
  public void twoJobsAtSameTime() {
    queue.add(new ReminderEntry(1, 1, 0));
    queue.add(new ReminderEntry(2, 1, 0));

    verify(workManager).scheduleNotification(1, true);

    assertEquals(2, queue.size());
  }

  @Test
  public void rescheduleForFirstJob() {
    queue.add(new ReminderEntry(1, 1, 0));

    verify(workManager).scheduleNotification(1, true);
  }

  @Test
  public void dontRescheduleForLaterJobs() {
    queue.add(new ReminderEntry(1, 1, 0));
    queue.add(new ReminderEntry(2, 2, 0));

    verify(workManager).scheduleNotification(1, true);
  }

  @Test
  public void rescheduleForNewerJob() {
    queue.add(new ReminderEntry(1, 2, 0));
    queue.add(new ReminderEntry(1, 1, 0));

    InOrder order = inOrder(workManager);
    order.verify(workManager).scheduleNotification(2, true);
    order.verify(workManager).scheduleNotification(1, true);
  }

  @Test
  public void rescheduleWhenCancelingOnlyJob() {
    queue.add(new ReminderEntry(1, 2, 0));
    queue.cancelReminder(1);

    InOrder order = inOrder(workManager);
    order.verify(workManager).scheduleNotification(2, true);
    order.verify(workManager).cancelNotifications();
  }

  @Test
  public void rescheduleWhenCancelingFirstJob() {
    queue.add(new ReminderEntry(1, 1, 0));
    queue.add(new ReminderEntry(2, 2, 0));

    queue.cancelReminder(1);

    InOrder order = inOrder(workManager);
    order.verify(workManager).scheduleNotification(1, true);
    order.verify(workManager).scheduleNotification(2, true);
  }

  @Test
  public void dontRescheduleWhenCancelingLaterJob() {
    queue.add(new ReminderEntry(1, 1, 0));
    queue.add(new ReminderEntry(2, 2, 0));

    queue.cancelReminder(2);

    verify(workManager).scheduleNotification(1, true);
  }

  @Test
  public void nextScheduledTimeIsZeroWhenQueueIsEmpty() {
    when(preferences.adjustForQuietHours(anyLong())).thenReturn(1234L);

    assertEquals(0, queue.nextScheduledTime());
  }

  @Test
  public void adjustNextScheduledTimeForQuietHours() {
    when(preferences.adjustForQuietHours(anyLong())).thenReturn(1234L);
    queue.add(new ReminderEntry(1, 1, 1));

    verify(workManager).scheduleNotification(1234, true);
  }

  @Test
  public void overdueJobsAreReturned() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(
                    singletonList(new ReminderEntry(1, now, TYPE_DUE)), queue.getOverdueJobs());
              }
            });
  }

  @Test
  public void twoOverdueJobsAtSameTimeReturned() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new ReminderEntry(2, now, TYPE_DUE));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(
                    asList(
                        new ReminderEntry(1, now, TYPE_DUE), new ReminderEntry(2, now, TYPE_DUE)),
                    queue.getOverdueJobs());
              }
            });
  }

  @Test
  public void twoOverdueJobsAtDifferentTimes() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now + 2 * ONE_MINUTE)
        .thawAfter(
            new Snippet() {
              {
                assertEquals(
                    asList(
                        new ReminderEntry(1, now, TYPE_DUE),
                        new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE)),
                    queue.getOverdueJobs());
              }
            });
  }

  @Test
  public void overdueJobsAreRemoved() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                queue.remove(queue.getOverdueJobs());
              }
            });

    assertEquals(singletonList(new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE)), queue.getJobs());
  }

  @Test
  public void multipleOverduePeriodsLapsed() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.add(new ReminderEntry(2, now + ONE_MINUTE, TYPE_DUE));
    queue.add(new ReminderEntry(3, now + 2 * ONE_MINUTE, TYPE_DUE));

    verify(workManager).scheduleNotification(now, true);

    Freeze.freezeAt(now + ONE_MINUTE)
        .thawAfter(
            new Snippet() {
              {
                queue.remove(queue.getOverdueJobs());
              }
            });

    assertEquals(
        singletonList(new ReminderEntry(3, now + 2 * ONE_MINUTE, TYPE_DUE)), queue.getJobs());
  }

  @Test
  public void clearShouldCancelExisting() {
    queue.add(new ReminderEntry(1, 1, 0));

    queue.clear();

    InOrder order = inOrder(workManager);
    order.verify(workManager).scheduleNotification(1, true);
    order.verify(workManager).cancelNotifications();
    assertEquals(0, queue.size());
  }

  @Test
  public void ignoreInvalidCancel() {
    long now = currentTimeMillis();

    queue.add(new ReminderEntry(1, now, TYPE_DUE));
    queue.cancelReminder(2);

    verify(workManager).scheduleNotification(now, true);
  }

  @Test
  public void allDuringSameMinuteAreOverdue() {
    DateTime now = new DateTime(2017, 9, 3, 0, 14, 6, 455);
    DateTime due = new DateTime(2017, 9, 3, 0, 14, 0, 0);
    DateTime snooze = new DateTime(2017, 9, 3, 0, 14, 59, 999);

    queue.add(new ReminderEntry(1, due.getMillis(), TYPE_DUE));
    queue.add(new ReminderEntry(2, snooze.getMillis(), TYPE_SNOOZE));
    queue.add(new ReminderEntry(3, due.plusMinutes(1).getMillis(), TYPE_DUE));

    verify(workManager).scheduleNotification(due.getMillis(), true);

    Freeze.freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                List<? extends NotificationQueueEntry> overdueJobs = queue.getOverdueJobs();
                assertEquals(
                    asList(
                        new ReminderEntry(1, due.getMillis(), TYPE_DUE),
                        new ReminderEntry(2, snooze.getMillis(), TYPE_SNOOZE)),
                    overdueJobs);
                queue.remove(overdueJobs);
                assertEquals(
                    singletonList(new ReminderEntry(3, due.plusMinutes(1).getMillis(), TYPE_DUE)),
                    queue.getJobs());
              }
            });
  }
}

File: app/src/test/java/org/tasks/jobs/NotificationQueueTest.kt
package org.tasks.jobs

import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.AdditionalAnswers
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.tasks.Freeze.Companion.freezeAt
import org.tasks.data.Alarm.Companion.TYPE_DATE_TIME
import org.tasks.data.Alarm.Companion.TYPE_SNOOZE
import org.tasks.preferences.Preferences
import org.tasks.time.DateTime
import org.tasks.time.DateTimeUtils
import java.util.concurrent.TimeUnit

class NotificationQueueTest {
    private lateinit var queue: NotificationQueue
    private lateinit var workManager: WorkManager
    private lateinit var preferences: Preferences

    @Before
    fun before() {
        preferences = Mockito.mock(Preferences::class.java)
        Mockito.`when`(preferences.adjustForQuietHours(ArgumentMatchers.anyLong()))
            .then(AdditionalAnswers.returnsFirstArg<Any>())
        workManager = Mockito.mock(WorkManager::class.java)
        queue = NotificationQueue(preferences, workManager)
    }

    @After
    fun after() {
        Mockito.verifyNoMoreInteractions(workManager)
    }

    @Test
    fun removeAlarmDoesntAffectOtherAlarm() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        queue.add(AlarmEntry(2, 2, now, TYPE_DATE_TIME))
        queue.remove(listOf(AlarmEntry(1, 1, now, TYPE_DATE_TIME)))
        freezeAt(now) {
            assertEquals(
                listOf(AlarmEntry(2, 2, now, TYPE_DATE_TIME)),
                queue.overdueJobs
            )
        }
    }

    @Test
    fun removeByTaskDoesntAffectOtherAlarm() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        queue.add(AlarmEntry(2, 2, now, TYPE_DATE_TIME))
        queue.cancelForTask(1)
        freezeAt(now) {
            assertEquals(
                listOf(AlarmEntry(2, 2, now, TYPE_DATE_TIME)),
                queue.overdueJobs
            )
        }
    }

    @Test
    fun rescheduleForFirstJob() {
        queue.add(AlarmEntry(1, 2, 3, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(3)
    }

    @Test
    fun dontRescheduleForLaterJobs() {
        queue.add(AlarmEntry(1, 2, 3, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 3, 4, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(3)
    }

    @Test
    fun rescheduleForNewerJob() {
        queue.add(AlarmEntry(1, 1, 2, TYPE_DATE_TIME))
        queue.add(AlarmEntry(1, 1, 1, TYPE_DATE_TIME))
        val order = Mockito.inOrder(workManager)
        order.verify(workManager).scheduleNotification(2)
        order.verify(workManager).scheduleNotification(1)
    }

    @Test
    fun rescheduleWhenCancelingOnlyJob() {
        queue.add(AlarmEntry(1, 1, 2, TYPE_DATE_TIME))
        queue.cancelForTask(1)
        val order = Mockito.inOrder(workManager)
        order.verify(workManager).scheduleNotification(2)
        order.verify(workManager).cancelNotifications()
    }

    @Test
    fun rescheduleWhenCancelingFirstJob() {
        queue.add(AlarmEntry(1, 1, 1, 0))
        queue.add(AlarmEntry(2, 2, 2, 0))
        queue.cancelForTask(1)
        val order = Mockito.inOrder(workManager)
        order.verify(workManager).scheduleNotification(1)
        order.verify(workManager).scheduleNotification(2)
    }

    @Test
    fun dontRescheduleWhenCancelingLaterJob() {
        queue.add(AlarmEntry(1, 1, 1, 0))
        queue.add(AlarmEntry(2, 2, 2, 0))
        queue.cancelForTask(2)
        Mockito.verify(workManager).scheduleNotification(1)
    }

    @Test
    fun nextScheduledTimeIsZeroWhenQueueIsEmpty() {
        Mockito.`when`(preferences.adjustForQuietHours(ArgumentMatchers.anyLong()))
            .thenReturn(1234L)
        assertEquals(0, queue.nextScheduledTime())
    }

    @Test
    fun adjustNextScheduledTimeForQuietHours() {
        Mockito.`when`(preferences.adjustForQuietHours(ArgumentMatchers.anyLong()))
            .thenReturn(1234L)
        queue.add(AlarmEntry(1, 1, 1, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(1234)
    }

    @Test
    fun overdueJobsAreReturned() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 1, now + ONE_MINUTE, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        freezeAt(now) {
            assertEquals(
                listOf(AlarmEntry(1, 1, now, TYPE_DATE_TIME)), queue.overdueJobs
            )
        }
    }

    @Test
    fun twoOverdueJobsAtSameTimeReturned() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 2, now, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        freezeAt(now) {
            assertEquals(
                setOf(
                    AlarmEntry(1, 1, now, TYPE_DATE_TIME),
                    AlarmEntry(2, 2, now, TYPE_DATE_TIME)
                ),
                queue.overdueJobs.toSet()
            )
        }
    }

    @Test
    fun twoOverdueJobsAtDifferentTimes() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 2, now + ONE_MINUTE, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        freezeAt(now + 2 * ONE_MINUTE) {
            assertEquals(
                listOf(
                    AlarmEntry(1, 1, now, TYPE_DATE_TIME),
                    AlarmEntry(2, 2, now + ONE_MINUTE, TYPE_DATE_TIME)
                ),
                queue.overdueJobs
            )
        }
    }

    @Test
    fun overdueJobsAreRemoved() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 2, now + ONE_MINUTE, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        freezeAt(now) {
            queue.remove(queue.overdueJobs)
        }
        assertEquals(
            listOf(AlarmEntry(2, 2, now + ONE_MINUTE, TYPE_DATE_TIME)), queue.getJobs()
        )
    }

    @Test
    fun multipleOverduePeriodsLapsed() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 2, now + ONE_MINUTE, TYPE_DATE_TIME))
        queue.add(AlarmEntry(3, 3, now + 2 * ONE_MINUTE, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(now)
        freezeAt(now + ONE_MINUTE) {
            queue.remove(queue.overdueJobs)
        }
        assertEquals(
            listOf(AlarmEntry(3, 3, now + 2 * ONE_MINUTE, TYPE_DATE_TIME)), queue.getJobs()
        )
    }


    @Test
    fun clearShouldCancelExisting() {
        queue.add(AlarmEntry(1, 1, 1, 0))
        queue.clear()
        val order = Mockito.inOrder(workManager)
        order.verify(workManager).scheduleNotification(1)
        order.verify(workManager).cancelNotifications()
        assertEquals(0, queue.size())
    }

    @Test
    fun ignoreInvalidCancelForByAlarm() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.remove(listOf(AlarmEntry(2, 2, now, TYPE_DATE_TIME)))
        Mockito.verify(workManager).scheduleNotification(now)
    }

    @Test
    fun ignoreInvalidCancelForTask() {
        val now = DateTimeUtils.currentTimeMillis()
        queue.add(AlarmEntry(1, 1, now, TYPE_DATE_TIME))
        queue.cancelForTask(2)
        Mockito.verify(workManager).scheduleNotification(now)
    }

    @Test
    fun allDuringSameMinuteAreOverdue() {
        val now = DateTime(2017, 9, 3, 0, 14, 6, 455)
        val due = DateTime(2017, 9, 3, 0, 14, 0, 0)
        val snooze = DateTime(2017, 9, 3, 0, 14, 59, 999)
        queue.add(AlarmEntry(1, 1, due.millis, TYPE_DATE_TIME))
        queue.add(AlarmEntry(2, 2, snooze.millis, TYPE_SNOOZE))
        queue.add(AlarmEntry(3, 3, due.plusMinutes(1).millis, TYPE_DATE_TIME))
        Mockito.verify(workManager).scheduleNotification(due.millis)
        freezeAt(now) {
            val overdueJobs = queue.overdueJobs
            assertEquals(
                listOf(
                    AlarmEntry(1, 1, due.millis, TYPE_DATE_TIME),
                    AlarmEntry(2, 2, snooze.millis, TYPE_SNOOZE)
                ),
                overdueJobs
            )
            queue.remove(overdueJobs)
            assertEquals(
                listOf(AlarmEntry(3, 3, due.plusMinutes(1).millis, TYPE_DATE_TIME)),
                queue.getJobs()
            )
        }
    }

    companion object {
        private val ONE_MINUTE = TimeUnit.MINUTES.toMillis(1)
    }
}